### PR TITLE
[Feature]: Official MAE

### PR DESCRIPTION
**Pull request by YuanLiuuuuuu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Official MAE.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/benchmarks/classification/_base_/models/vit_base_16.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+model = dict(
```
Comment (fangyixiao18) [inline configs/benchmarks/classification/_base_/models/vit_base_16.py]: I think using the config file name format like `vit-base-p16` is better, the same name as mmcls.

File: configs/benchmarks/classification/_base_/models/vit_base_16_linear_eval.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+model = dict(
+    type='MAELinearClassification',
+    backbone=dict(
+        type='MAEClsViT',
+        arch='b',
+        patch_size=16,
+        global_pool=True,
+        finetune=False,
+        final_norm=False),
+    head=dict(type='MAELinearEvalHead', num_classes=1000, embed_dim=768))
```
Comment (fangyixiao18) [inline configs/benchmarks/classification/_base_/models/vit_base_16_linear_eval.py]: Name consistency? `linear_eval`, `MAELinearClassification` and `MAELinearEvalHead` are not consistent, can we name them with the same convention?
Besides, I think the `vit-base-p16_linear-eval.py` format is better.

File: configs/selfsup/mae/README.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# MAE Experimental
+
```
Comment (fangyixiao18) [inline configs/selfsup/mae/README.md:2]: Missing the paper title and url. 
Please check the format here https://github.com/open-mmlab/mmselfsup/pull/177

File: configs/selfsup/mae/README.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+meaningful self-supervisory task. Coupling these two designs enables us to
+train large models efficiently and effectively: we accelerate
+training (by 3× or more) and improve accuracy. Our scalable approach allows
+for learning high-capacity models that generalize well: e.g., a vanilla
+ViT-Huge model achieves the best accuracy (87.8%) among
+methods that use only ImageNet-1K data. Transfer performance in downstream tasks outperforms supervised pretraining and shows promising scaling behavior.
+
+<!-- [IMAGE] -->
+<div align="center">
+<img src="../../../resources/model_zoo/mae.png" width="40%"/>
```
Comment (fangyixiao18) [inline configs/selfsup/mae/README.md]: it is no need to make a folder to save images. we can use the link.

File: mmselfsup/datasets/builder.py [hunk 5]
```diff
@@ -53,7 +53,7 @@ def build_dataloader(dataset,
                      replace=False,
                      seed=None,
                      pin_memory=True,
-                     persistent_workers=True,
+                     persistent_workers=False,
```
Comment (fangyixiao18) [inline mmselfsup/datasets/builder.py]: why do you change this?

File: mmselfsup/models/algorithms/mae.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+from ..builder import ALGORITHMS, build_backbone, build_head, build_neck
+from ..utils import get_2d_sincos_pos_embed
+from .base import BaseModel
+
+
+@ALGORITHMS.register_module('MAE')
+class MAE(BaseModel):
+    """MAE.
+
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/mae.py:9]: missing the paper name and url.

File: mmselfsup/models/algorithms/mae.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    backbone (dict): Config dict for encoder.     Defaults to None. neck
+    (dict): Config dict for encoder.     Defaults to None head (dict): Config
+    dict for loss functions.     Defaults to None. init_cfg (dict): Config dict
+    for weight initialization.     Defaults to None.
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/mae.py]: check the docstring format

File: mmselfsup/datasets/pipelines/transforms.py [hunk 8]
```diff
@@ -5,6 +5,7 @@
 import torch
 from mmcv.utils import build_from_cfg
 from PIL import Image, ImageFilter
+from timm.data import create_transform
```
Comment (fangyixiao18) [inline mmselfsup/datasets/pipelines/transforms.py:8]: can we write our own function instead of importing timm?

File: mmselfsup/models/algorithms/mae_linprobe.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/mae_linprobe.py]: 'mae_linprobe.py' and 'mae_finetune.py' can be merged, two files are kind of redundant.

File: tools/slurm_train.sh [hunk 10]
```diff
@@ -6,9 +6,9 @@ PARTITION=$1
 JOB_NAME=$2
 CONFIG=$3
 WORK_DIR=$4
-GPUS=${GPUS:-8}
+GPUS=${GPUS:-32}
 GPUS_PER_NODE=${GPUS_PER_NODE:-8}
-CPUS_PER_TASK=${CPUS_PER_TASK:-5}
+CPUS_PER_TASK=${CPUS_PER_TASK:-1}
```
Comment (fangyixiao18) [inline tools/slurm_train.sh]: I think it is no need to change these two default value

File: tools/train.py [hunk 11]
```diff
@@ -133,7 +133,6 @@ def main():
     cfg.seed = seed
     meta['seed'] = seed
     meta['exp_name'] = osp.basename(args.config)
-
```
Comment (fangyixiao18) [inline tools/train.py]: A blank line to make it more clear to build the model.

File: mmselfsup/models/algorithms/mae_linprobe.py (hunk 9 above)
Comment (fangyixiao18) [reply mmselfsup/models/algorithms/mae_linprobe.py]: Or modifying the original classification.py to avoid the code redundancy.

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]
No UT codes?

File: configs/benchmarks/classification/_base_/models/vit_base_16.py (hunk 1 above)
Comment (YuanLiuuuuuu) [reply configs/benchmarks/classification/_base_/models/vit_base_16.py]: Thank you for your valuable feedbacks. Your suggestion is reasonable and I will fix it.

File: configs/benchmarks/classification/_base_/models/vit_base_16_linear_eval.py (hunk 2 above)
Comment (YuanLiuuuuuu) [reply configs/benchmarks/classification/_base_/models/vit_base_16_linear_eval.py]: Thank you for your valuable feedbacks. Your suggestion is reasonable and I will fix it.

File: mmselfsup/datasets/builder.py (hunk 5 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/datasets/builder.py]: Thank you for your valuable reviews. For the convenience of debugging, I change the default value of `persistent_worker` and forgot to change it back in my official PR. I will fix it~

File: tools/slurm_train.sh (hunk 10 above)
Comment (YuanLiuuuuuu) [reply tools/slurm_train.sh]: Thank you for your valuable reviews. For the convenience of debugging, I change the default value of `CPUS_PER_TASK` and forgot to change it back in my official PR. I will fix it~

File: configs/benchmarks/classification/_base_/models/vit-base-p16._ft.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
+model = dict(
```
Comment (fangyixiao18) [inline configs/benchmarks/classification/_base_/models/vit-base-p16._ft.py]: there is a `.` before `_ft` in file name.

File: mmselfsup/models/backbones/mae_pretrain_vit.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        pass
+
+    def random_masking(self, x, mask_ratio=0.75):
+        """Generate the mask for MAE Pre-training.
+
+        Args:
+            x (torch.tensor): Image with data augmentation applied.
+            mask_ratio (float): The mask ratio of total patches.
+                Defaults to 0.75.
+        """
```
Comment (fangyixiao18) [inline mmselfsup/models/backbones/mae_pretrain_vit.py:101]: I think it is better to add `Returns` to the doc string for easy understanding.

File: mmselfsup/models/heads/mae_finetune_head.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+import torch.nn.functional as F
+from mmcv.cnn.utils.weight_init import trunc_normal_
+from mmcv.runner import BaseModule
+from torch import nn
+
+from ..builder import HEADS
+
+
+class SoftTargetCrossEntropy(nn.Module):
+
```
Comment (fangyixiao18) [inline mmselfsup/models/heads/mae_finetune_head.py]: I think it is better to add some doc string here

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 15]
```diff
@@ -121,6 +121,10 @@ tensorboard.sh
 replace.sh
 benchmarks/detection/datasets
 benchmarks/detection/output
+ftest*
```
Comment (ZwwWayne) [inline .gitignore]: We should not involve unnecessary changes in this PR, which supports MAE. This distracts the review and should be put in another PR if necessary.

File: mmselfsup/models/algorithms/mae_classification.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+            img (Tensor): Input images of shape (N, C, H, W).
+                Typically these should be mean centered and std scaled.
+            label (Tensor): Ground-truth labels.
+            kwargs: Any keyword arguments to be used to forward.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        if self.finetune:
+            img, label = self.mix_up(img, label)
```
Comment (ZwwWayne) [inline mmselfsup/models/algorithms/mae_classification.py]: The usage of mix-up should not be related to self.finetune. Use a specific flag (e.g., use_mixup) to control whether to use mix_up.

File: mmselfsup/models/backbones/mae_cls_vit.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        # stole cls_tokens impl from Phil Wang, thanks
+        cls_tokens = self.cls_token.expand(B, -1, -1)
+        x = torch.cat((cls_tokens, x), dim=1)
+        x = x + self.pos_embed
+        x = self.drop_after_pos(x)
+
+        for i, layer in enumerate(self.layers):
+            x = layer(x)
+
+            if i == len(self.layers) - 1 and self.final_norm:
```
Comment (ZwwWayne) [inline mmselfsup/models/backbones/mae_cls_vit.py]: 1. line 106-114 is confusing, is `self.final_norm=True and self.global_pool=True` expected? It seems that `self.final_norm=True and self.global_pool=False` will go to line 113.
2. Considering move global_pool to neck so that we can reuse backbone in mmcls.

File: mmselfsup/models/algorithms/mae_classification.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from ..builder import ALGORITHMS, build_backbone, build_head
+from ..utils import Mixup
+from .base import BaseModel
+
+
+@ALGORITHMS.register_module()
+class MAEClassification(BaseModel):
```
Comment (ZwwWayne) [inline mmselfsup/models/algorithms/mae_classification.py]: The main modification is about use mixup? Consider merge mae_classification with existing classification algorithm by supporting mixup in existing algorithm. This backbone seems to be not specially designed for MAE for now.

File: mmselfsup/models/heads/mae_linprobe_head.py [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+        embed_dim (int): The dim of the feature before the classifier head.
+        num_classes (int): The total classes. Defaults to 1000.
+    """
+
+    def __init__(self, embed_dim, num_classes=1000):
+        super(MAELinprobeHead, self).__init__()
+        self.head = nn.Linear(embed_dim, num_classes)
+        self.bn = nn.BatchNorm1d(embed_dim, affine=False, eps=1e-6)
+        self.criterion = nn.CrossEntropyLoss()
+        nn.init.constant_(self.head.bias, 0)
```
Comment (ZwwWayne) [inline mmselfsup/models/heads/mae_linprobe_head.py:25]: should do that in init_weights()

File: mmselfsup/models/heads/mae_pretrain_head.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+
+    Args:
+        norm_pix_loss (bool): Whether or not normalize target.
+            Defaults to False.
+        patch_size (int): Patch size. Defaults to 16.
+    """
+
+    def __init__(self, norm_pix_loss=False, patch_size=16):
+        super(MAEPretrainHead, self).__init__()
+        self.norm_pix_loss = norm_pix_loss
```
Comment (ZwwWayne) [inline mmselfsup/models/heads/mae_pretrain_head.py]: norm_pix_loss -> norm_pix

File: mmselfsup/models/utils/mixup.py [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+        yl, yu, xl, xu = rand_bbox_minmax(img_shape, ratio_minmax, count=count)
+    else:
+        yl, yu, xl, xu = rand_bbox(img_shape, lam, count=count)
+    if correct_lam or ratio_minmax is not None:
+        bbox_area = (yu - yl) * (xu - xl)
+        lam = 1. - bbox_area / float(img_shape[-2] * img_shape[-1])
+    return (yl, yu, xl, xu), lam
+
+
+class Mixup:
```
Comment (ZwwWayne) [inline mmselfsup/models/utils/mixup.py:98]: Rethinking its relationship with the MixUp module in mmcls

File: mmselfsup/models/utils/position_embedding.py [hunk 22]
```diff
@@ -0,0 +1,52 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import numpy as np
+
+
+def get_2d_sincos_pos_embed(embed_dim, grid_size, cls_token=False):
```
Comment (ZwwWayne) [inline mmselfsup/models/utils/position_embedding.py]: Why do we do it by numpy rather than torch? Can we unify it for simplicity or efficiency?

File: tools/benchmarks/classification/dist_train_linear.sh [hunk 23]
```diff
@@ -8,11 +8,19 @@ PRETRAIN=$2  # pretrained model
...
+NNODES=${NNODES:-1}
+NODE_RANK=${NODE_RANK:-0}
+MASTER_ADDR=${MASTER_ADDR:-'127.0.0.1'}
 
 # set work_dir according to config path and pretrained model to distinguish different models
 WORK_DIR="$(echo ${CFG%.*} | sed -e "s/configs/work_dirs/g")/$(echo $PRETRAIN | rev | cut -d/ -f 1 | rev)"
 
-python -m torch.distributed.launch --nproc_per_node=$GPUS --master_port=$PORT \
+python -m torch.distributed.launch \
+    --nproc_per_node=$GPUS \
```
Comment (ZwwWayne) [inline tools/benchmarks/classification/dist_train_linear.sh]: Such a modification should be updated in another PR, which does not related to MAE.

File: .gitignore (hunk 15 above)
Comment (YuanLiuuuuuu) [reply .gitignore]: Thank you for your valuable comments.  My experiment will generate some temporal files, but I do not expect git to track it. So I add it to `gitignore` on my local machine, and forgot to delete it before I push it to remote. I will fix it as soon as possible.

File: mmselfsup/models/algorithms/mae_classification.py (hunk 16 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/algorithms/mae_classification.py]: Thank you for your valuable comments. Your suggestion is reasonable, and I will fix it.

File: mmselfsup/models/heads/mae_pretrain_head.py (hunk 20 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/heads/mae_pretrain_head.py]: Thank you for your valuable comments. `norm_pix` seems more intuitive.

File: tools/benchmarks/classification/dist_train_linear.sh (hunk 23 above)
Comment (YuanLiuuuuuu) [reply tools/benchmarks/classification/dist_train_linear.sh]: Thank you for your valuable comments. The same problem as the modification in `gitignore` and I will fix it as soon as possible.

File: mmselfsup/models/backbones/mae_cls_vit.py (hunk 17 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/backbones/mae_cls_vit.py]: Thank you for your valuable comments.

1. Your idea is right. I can only use `self.final_norm` in `forward` to control the whole logical flow.
2. We can not directly use the backbone from mmcls, even without `self.global_pool`. Because we need our customized `train` to freeze the parameters during linear probing.

File: mmselfsup/models/heads/mae_linprobe_head.py (hunk 19 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/heads/mae_linprobe_head.py:25]: Thank you for your valuable comments. I will fix it as soon as possible.

File: mmselfsup/models/utils/position_embedding.py (hunk 22 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/utils/position_embedding.py]: Thank you for your valuable comments. I use torch to construct the pe, instead of numpy, in my latest PR.

File: mmselfsup/models/heads/mae_finetune_head.py [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+from mmcls.models import LabelSmoothLoss
+from mmcv.cnn.utils.weight_init import trunc_normal_
+from mmcv.runner import BaseModule
+from torch import nn
+
+from ..builder import HEADS
+
+
+@HEADS.register_module()
+class MAEFinetuneHead(BaseModule):
```
Comment (Jiahao000) [inline mmselfsup/models/heads/mae_finetune_head.py:11]: Head is currently redundant. Can `mae_pretrain_head.py`, `mae_finetune_head.py`, `mae_linearprob_head.py` be merged into one `mae_head.py` file?

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+# model
+model = dict(backbone=dict(init_cfg=dict(prefix='backbone.')))
+
+# optimizer
+optimizer = dict(
+    lr=1e-3 * 1024 / 256,
+    paramwise_options={
+        'norm': dict(weight_decay=0.),
+        'bias': dict(weight_decay=0.),
+        'pos_embed': dict(weight_decay=0.003697205891018715),
```
Comment (Jiahao000) [inline configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py]: Why use these specific numbers here?

File: mmselfsup/models/backbones/mae_cls_vit.py [hunk 26]
```diff
@@ -0,0 +1,110 @@
+import torch
```
Comment (Jiahao000) [inline mmselfsup/models/backbones/mae_cls_vit.py]: Can `mae_cls_vit.py` and `mae_pretrain_vit.py` be merged into one file? Also, can `MAEClsViT` be written as a more general form so that downstream linear probing and fine-tuning can be equally applicable to other methods with the same backbone.

**Review by Jiahao000 — CHANGES_REQUESTED** [review]

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py (hunk 25 above)
Comment (ZwwWayne) [reply configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py]: we should leave comments here to indicate the reason

File: mmselfsup/models/algorithms/mae.py [hunk 27]
```diff
@@ -0,0 +1,90 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from torch import nn
+
+from ..builder import ALGORITHMS, build_backbone, build_head, build_neck
+from ..utils import build_2d_sincos_position_embedding
+from .base import BaseModel
+
+
+@ALGORITHMS.register_module('MAE')
```
Comment (ZwwWayne) [inline mmselfsup/models/algorithms/mae.py]: @ALGORITHMS.register_module('MAE') -> @ALGORITHMS.register_module() is equally effective

File: mmselfsup/models/algorithms/mae.py [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+    def __init__(self, backbone=None, neck=None, head=None, init_cfg=None):
+        super(MAE, self).__init__(init_cfg)
+        assert backbone is not None
+        self.backbone = build_backbone(backbone)
+        assert neck is not None
+        self.neck = build_neck(neck)
+        self.neck.num_patches = self.backbone.patch_embed.num_patches
+        assert head is not None
+        self.head = build_head(head)
+        self.initialize_weights()
```
Comment (ZwwWayne) [inline mmselfsup/models/algorithms/mae.py]: should merge self.initialize_weights and self._init_weights

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+        'head': dict(lr_mult=1.0)
+    })
+
+# learning policy
+lr_config = dict(
+    policy='CosineAnnealing',
+    min_lr=1e-6,
+    warmup='linear',
+    warmup_iters=5,
+    warmup_ratio=1e-4,
```
Comment (ZwwWayne) [inline configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py:56]: inconsistent with the official code, should use warmup_ratio=1/5 and by_epoch=True

File: configs/selfsup/mae/mae_vit-b-p16_8xb512-coslr-400e_in1k.py [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+        'pos_embed': dict(weight_decay=0.),
+        'mask_token': dict(weight_decay=0.),
+        'cls_token': dict(weight_decay=0.)
+    })
+optimizer_config = dict()
+
+# learning policy
+lr_config = dict(
+    policy='CosineAnnealing',
+    min_lr=1e-5,
```
Comment (ZwwWayne) [inline configs/selfsup/mae/mae_vit-b-p16_8xb512-coslr-400e_in1k.py]: inconsistent with the official code, should be min_lr=0, warmup_ratio=1/40, by_epoch=True to be consistent.

Comment (ZwwWayne) [reply configs/selfsup/mae/mae_vit-b-p16_8xb512-coslr-400e_in1k.py]: The learning rate schedule is still a bit inconsistent, which is warmup + cosine annealing rather than warmup * cosine implemented in mmcv.

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py (hunk 25 above)
Comment (YuanLiuuuuuu) [reply configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py]: Thank you for your valuable comments. During fine-tuning, we will use layer-wise learning rate scheduling, with a base `lr` of 0.65.

File: mmselfsup/models/heads/mae_finetune_head.py (hunk 24 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/heads/mae_finetune_head.py:11]: Thank you for your valuable comments. Since `mae_pretrain_head.py`, `mae_finetune_head.py` and `mae_linprobe_head` share different `forward` function, it's not convenient to merge them into one fun. If we merge them, we have to introduce extra params to control the `workflow` and the `forward` func will become too complex.

File: mmselfsup/models/backbones/mae_cls_vit.py (hunk 26 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/backbones/mae_cls_vit.py]: Thank you for your valuable comments.  It's not convenient to merge `mae_cls_vit.py` and `mae_pretrain_vit.py` into one file, since they share different `forward` func. For `MAEClsViT`, it has specific initialization config, so it's not appropriate to rewrite it in a more general form.

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+model = dict(backbone=dict(init_cfg=dict(prefix='backbone.')))
+
+# optimizer
+optimizer = dict(
+    lr=1e-3 * 1024 / 256,
+    paramwise_options={
+        'norm': dict(weight_decay=0.),
+        'bias': dict(weight_decay=0.),
+        'pos_embed': dict(weight_decay=0.003697205891018715),
+        'cls_token': dict(weight_decay=0.003697205891018715),
```
Comment (ZwwWayne) [inline configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py]: This part is inconsistent with the official code. cls_token's lr should be decayed, its weight decay should be 0?

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py (hunk 25 above)
Comment (ZwwWayne) [inline configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py]: It seems that the weight_decay and lr_mult uses similar value, should be a mistake I guess.

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py (hunk 31 above)
Comment (ZwwWayne) [reply configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py]: We should inherit a OptimizerConstructor to do this automatically rather than fill the number manually

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py (hunk 25 above)
Comment (YuanLiuuuuuu) [reply configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py]: Thank you for your valuable comments. Yes, it's a mistake. There is something wrong with making this commit, but my result was obtained by using the correct config.

File: mmselfsup/models/algorithms/mae.py (hunk 27 above)
Comment (YuanLiuuuuuu) [reply mmselfsup/models/algorithms/mae.py]: Thank you for your valuable comments. I changed the original name of this algorithm to `MAE`, but forgot to delete `MAE`  when registering this  algorithm.

File: mmselfsup/models/heads/mae_finetune_head.py (hunk 24 above)
Comment (Jiahao000) [reply mmselfsup/models/heads/mae_finetune_head.py:11]: If this is the case, I suggest writing these three classes into one `mae_head.py` file instead of three separate files.

File: configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py (hunk 25 above)
Comment (Jiahao000) [reply configs/benchmarks/classification/imagenet/vit-b-p16_ft-8xb128-coslr-100e_in1k.py]: Better to add comments here for the specific `0.003697205891018715`

File: mmselfsup/models/heads/mae_finetune_head.py (hunk 24 above)
Comment (fangyixiao18) [reply mmselfsup/models/heads/mae_finetune_head.py:11]: Or I think maybe it is better to merge finetune and linprobe heads and leave the pretrain head

File: configs/selfsup/mae/README.md [hunk 32]
```diff
@@ -0,0 +1,58 @@
+# MAE
+
+[Masked Autoencoders Are Scalable Vision Learners](https://arxiv.org/abs/2111.06377)
```
Comment (fangyixiao18) [inline configs/selfsup/mae/README.md]: add `> ` at the begining of the line

File: configs/selfsup/mae/README.md [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+
+
```
Comment (fangyixiao18) [inline configs/selfsup/mae/README.md:6]: add `<!-- [ALGORITHM] -->` here

File: configs/selfsup/mae/README.md [hunk 34]
```diff
@@ -0,0 +1,58 @@
...
+## Abstract
+
+<!-- [ABSTRACT] -->
```
Comment (fangyixiao18) [inline configs/selfsup/mae/README.md]: remove this

File: configs/selfsup/mae/README.md [hunk 35]
```diff
@@ -0,0 +1,58 @@
...
+and mask tokens. Second, we find that masking a high proportion
+of the input image, e.g., 75%, yields a nontrivial and
+meaningful self-supervisory task. Coupling these two designs enables us to
+train large models efficiently and effectively: we accelerate
+training (by 3× or more) and improve accuracy. Our scalable approach allows
+for learning high-capacity models that generalize well: e.g., a vanilla
+ViT-Huge model achieves the best accuracy (87.8%) among
+methods that use only ImageNet-1K data. Transfer performance in downstream tasks outperforms supervised pretraining and shows promising scaling behavior.
+
+<!-- [IMAGE] -->
```
Comment (fangyixiao18) [inline configs/selfsup/mae/README.md]: remove this

File: configs/selfsup/mae/README.md [hunk 36]
```diff
@@ -0,0 +1,58 @@
...
+| :------: | :-------------: | :---------------: | :------------------------------------------------: | :--------------------------------------------------------------------------------------------: | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
+| ViT-B/16 |       400       |       82.93       | [config](./mae_vit-b-16_8xb512-coslr-400e_in1k.py) | [config](../../benchmarks/classification/imagenet/vit-b-16_8xb128-coslr-100e-finetune_in1k.py) | [model](https://download.openmmlab.com/mmselfsup/mae/mae_vit-b-16_8xb512-coslr-400e_in1k_20220107-4f3b0ae2.pth) &#124; [log](https://github.com/open-mmlab/mmselfsup/blob/feature/mae_experimental/configs/selfsup/mae/mae_vit-b-16_8xb512-coslr-400e_in1k.py) |
+
+
+## Citation
+
+<!-- [ALGORITHM] -->
```
Comment (fangyixiao18) [inline configs/selfsup/mae/README.md]: remove this
You can check this PR https://github.com/open-mmlab/mmselfsup/pull/206

File: mmselfsup/models/backbones/mae_pretrain_vit.py (hunk 13 above)
Comment (fangyixiao18) [reply mmselfsup/models/backbones/mae_pretrain_vit.py:101]: Missing `Returns`?

**Review by fangyixiao18 — CHANGES_REQUESTED** [review]
